rpc/namenode.go

```diff
@@ -109,19 +172,28 @@ func (c *NamenodeConnection) Execute(method string, req proto.Message, resp prot
 	defer c.reqLock.Unlock()
 
 	c.currentRequestID++
-	err := c.writeRequest(method, req)
+
+R:
+	err := c.resolveConnection()
 	if err != nil {
-		c.conn.Close()
 		return err
 	}
 
+	err = c.writeRequest(method, req)
+	if err != nil {
+		c.markFailure(err)
+		goto R
```

Comment:
Elsewhere in the code I use `for` loops and `continue`/`break` rather than goto. This is probably a bit shorter, but I think harder to read.